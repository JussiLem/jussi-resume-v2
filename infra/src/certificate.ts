import { aws_certificatemanager as cm, aws_route53 as route53, CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

type CertificateStackProps = {
  domainName: string;
} & StackProps;

export class CertificateStack extends Stack {
  constructor(scope: Construct, id: string, props: CertificateStackProps) {
    super(scope, id, props);
    const { domainName } = props;

    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
      domainName,
    });
    const certificate = new cm.Certificate(this, 'SiteCert', {
      domainName,
      subjectAlternativeNames: [`*.${domainName}`],
      validation: cm.CertificateValidation.fromDns(hostedZone),
    });
    new CfnOutput(this, 'CertificateArn', {
      value: certificate.certificateArn,
      exportName: 'resume-cert-arn',
    });
  }
}
