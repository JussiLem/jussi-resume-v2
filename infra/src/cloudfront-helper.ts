import { aws_cloudfront as cloudfront } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export const defaultCloudfrontFunction = (scope: Construct): cloudfront.Function => {
  // generate a stable unique id for the cloudfront function and use it
  // both for the function name and the logical id of the function so if
  // it is changed the function will be recreated.
  // see https://github.com/aws/aws-cdk/issues/15523
  const functionId = `SetHttpSecurityHeaders${scope.node.addr}`;

  return new cloudfront.Function(scope, 'SetHttpSecurityHeaders', {
    functionName: functionId,
    code: cloudfront.FunctionCode.fromInline(
      "function handler(event) { var response = event.response; \
      var headers = response.headers; \
      headers['strict-transport-security'] = { value: 'max-age=63072000; includeSubdomains; preload'}; \
      headers['content-security-policy'] = { value: \"default-src 'none'; img-src 'self'; script-src 'self'; style-src 'self'; manifest-src 'self'; connect-src 'self'; object-src 'none'\"}; \
      headers['x-content-type-options'] = { value: 'nosniff'}; \
      headers['x-frame-options'] = {value: 'DENY'}; \
      headers['x-xss-protection'] = {value: '1; mode=block'}; \
      return response; \
    }",
    ),
  });
};
