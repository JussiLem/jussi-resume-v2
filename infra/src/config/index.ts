// for development, use account/region from cdk cli
import { env } from '../common';
import { ResumeProps } from '../types';

const devEnv = {
  account: env('CDK_DEFAULT_ACCOUNT'),
  region: env('CDK_DEFAULT_REGION'),
};
const domainName = env('DOMAIN_NAME');
const certificateArn = env('CERTIFICATE_ARN');
const userPoolId = env('USER_POOL_ID');
const hostedZoneId = env('HOSTED_ZONE_ID');

export default { devEnv, domainName, certificateArn, userPoolId, hostedZoneId } as ResumeProps;
