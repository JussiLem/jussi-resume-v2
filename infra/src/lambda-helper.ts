import { Aws, aws_ec2 as ec2, aws_iam as iam, aws_lambda as lambda } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export const deployLambdaFunction = (
  scope: Construct,
  lambdaFunctionProps: lambda.FunctionProps,
  functionId?: string,
  vpc?: ec2.IVpc,
): lambda.Function => {
  const _functionId = functionId ? functionId : 'LambdaFunction';
  const _functionRoleId = _functionId + 'ServiceRole';
  if (vpc && lambdaFunctionProps.vpc) {
    throw Error('Cannot provide VPC both in props and function argument');
  }
  const lambdaServiceRole = new iam.Role(scope, _functionRoleId, {
    assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    inlinePolicies: {
      LambdaFunctionServiceRolePolicy: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            actions: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
            resources: [`arn:${Aws.PARTITION}:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:log-group:/aws/lambda/*`],
          }),
        ],
      }),
    },
  });

  if (lambdaFunctionProps.vpc || vpc) {
    lambdaServiceRole.addToPolicy(
      new iam.PolicyStatement({
        actions: [
          'ec2:CreateNetworkInterface',
          'ec2:DescribeNetworkInterfaces',
          'ec2:DeleteNetworkInterface',
          'ec2:AssignPrivateIpAddresses',
          'ec2:UnassignPrivateIpAddresses',
        ],
        resources: ['*'],
      }),
    );
  }
  return new lambda.Function(scope, _functionId, {
    ...lambdaFunctionProps,
    role: lambdaServiceRole,
  });
};
