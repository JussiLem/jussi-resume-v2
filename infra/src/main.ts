import { join } from 'path';
import {
  aws_route53 as route53,
  aws_certificatemanager as cm,
  App,
  Aspects,
  Stack,
  StackProps,
  CfnOutput,
} from 'aws-cdk-lib';
import { AwsSolutionsChecks } from 'cdk-nag';
import { Construct } from 'constructs';

const nextPath = join(__dirname, '../.next');
// const outputPath = join(nextPath, '/build');

export const env = (name: string): string => {
  const value = process.env[name];

  if (!value) {
    throw new Error(`Missing: process.env['${name}'].`);
  }

  return value;
};

type CertificateStackProps = {
  domainName: string;
} & StackProps;

export class CertificateStack extends Stack {
  constructor(scope: Construct, id: string, props: CertificateStackProps) {
    super(scope, id, props);
    const { domainName } = props;

    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
      domainName,
    });
    const certificate = new cm.Certificate(this, 'SiteCert', {
      domainName,
      subjectAlternativeNames: [`*.${domainName}`],
      validation: cm.CertificateValidation.fromDns(hostedZone),
    });
    new CfnOutput(this, 'CertificateArn', {
      value: certificate.certificateArn,
      exportName: 'resume-cert-arn',
    });
  }
}
type ResumeProps = {
  nextStaticDir: string;
} & StackProps;
export class Resume extends Stack {
  constructor(scope: Construct, id: string, props: ResumeProps) {
    super(scope, id, props);

    // cm.Certificate.fromCertificateArn(this, 'id', 'certArn');
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: env('CDK_DEFAULT_ACCOUNT'),
  region: env('CDK_DEFAULT_REGION'),
};

const app = new App();
Aspects.of(app).add(new AwsSolutionsChecks());

new CertificateStack(app, 'certificate', {
  env: {
    account: devEnv.account,
    region: 'us-east-1',
  },
  domainName: env('DOMAIN_NAME'),
});
new Resume(app, 'resume', { env: devEnv, nextStaticDir: nextPath });

app.synth();
