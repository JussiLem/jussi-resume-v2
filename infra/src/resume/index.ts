export * from './resume';
export * from './certificate';
export { NextJsServerless } from './nextJsServerless';
