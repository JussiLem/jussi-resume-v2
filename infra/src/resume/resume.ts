import {
  aws_cloudfront as cloudfront,
  aws_cloudfront_origins as origins,
  aws_certificatemanager as cm,
  aws_s3 as s3,
  aws_s3_deployment as deploy,
  Duration,
  Stack,
  StackProps,
} from 'aws-cdk-lib';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import { buildS3Bucket } from '../s3-bucket-helper';

type NextJsServerlessProps = {
  certificateArn: string;
};

export class NextJsServerless extends Construct {
  public readonly distribution: cloudfront.Distribution;
  public readonly cloudFrontFunction?: cloudfront.Function;
  // public readonly s3BucketInterface: s3.IBucket;
  public readonly s3Bucket?: s3.Bucket;

  constructor(scope: Construct, id: string, props: NextJsServerlessProps) {
    super(scope, id);
    this.s3Bucket = buildS3Bucket(this, undefined, 'PublicAssets');
    NagSuppressions.addResourceSuppressions(this.s3Bucket, [{ id: 'AwsSolutions-S1', reason: 'For money saving' }]);

    // new deploy.BucketDeployment(this, 'NoCacheFilesDeployment', {
    //   sources: [deploy.Source.asset('../.next/server')],
    //   destinationBucket: this.s3Bucket,
    //   cacheControl: [deploy.CacheControl.fromString('public, max-age=0, must-revalidate')],
    //   prune: false,
    // });

    new deploy.BucketDeployment(this, 'StaticFilesDeployment', {
      sources: [deploy.Source.asset('../out')],
      destinationBucket: this.s3Bucket,
      cacheControl: [deploy.CacheControl.fromString('public, max-age=31536000, immutable')],
      prune: false,
    });
    this.cloudFrontFunction = new cloudfront.Function(this, 'FormatRequestLambda', {
      comment: 'Formats path for S3',
      code: cloudfront.FunctionCode.fromFile({
        filePath: 'functions/format-request.js',
      }),
    });

    this.distribution = new cloudfront.Distribution(this, 'Distribution', {
      defaultBehavior: {
        origin: new origins.S3Origin(this.s3Bucket),
        allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
        viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
        cachePolicy: new cloudfront.CachePolicy(this, 'DefaultCachePolicy', {
          enableAcceptEncodingBrotli: true,
          enableAcceptEncodingGzip: true,
          minTtl: Duration.seconds(2),
          maxTtl: Duration.seconds(600),
          defaultTtl: Duration.seconds(2),
        }),
        responseHeadersPolicy: cloudfront.ResponseHeadersPolicy.SECURITY_HEADERS,
        originRequestPolicy: cloudfront.OriginRequestPolicy.CORS_S3_ORIGIN,
        // functionAssociations: [
        //   {
        //     eventType: cloudfront.FunctionEventType.VIEWER_REQUEST,
        //     function: this.cloudFrontFunction,
        //   },
        // ],
      },
      additionalBehaviors: {
        '_next/*': {
          origin: new origins.S3Origin(this.s3Bucket),
          allowedMethods: cloudfront.AllowedMethods.ALLOW_GET_HEAD_OPTIONS,
          viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
          cachePolicy: cloudfront.CachePolicy.CACHING_OPTIMIZED,
          responseHeadersPolicy: cloudfront.ResponseHeadersPolicy.SECURITY_HEADERS,
          originRequestPolicy: cloudfront.OriginRequestPolicy.CORS_S3_ORIGIN,
        },
      },
      priceClass: cloudfront.PriceClass.PRICE_CLASS_100,
      domainNames: ['jussilemmetyinen.fi'],
      certificate: cm.Certificate.fromCertificateArn(this, 'id', props.certificateArn),
    });
  }
}

type ResumeProps = {
  certificateArn: string;
} & StackProps;
export class Resume extends Stack {
  constructor(scope: Construct, id: string, props: ResumeProps) {
    super(scope, id, props);
    new NextJsServerless(this, 'NextJs', {
      certificateArn: props.certificateArn,
    });
    NagSuppressions.addStackSuppressions(this, [{ id: 'AwsSolutions-CFR3', reason: 'Save money' }]);
    NagSuppressions.addStackSuppressions(this, [{ id: 'AwsSolutions-IAM4', reason: 'Change later to own policy' }]);
    NagSuppressions.addStackSuppressions(this, [
      { id: 'AwsSolutions-IAM5', reason: 'Change later to stop using wildcard' },
    ]);

    // new nodejs.NodejsFunction(this, 'RegenerationLambda', {
    //   handler: 'index.handler',
    //   entry: path.join('.next', 'regeneration-lambda'),
    // });
    // cm.Certificate.fromCertificateArn(this, 'id', 'certArn');
  }
}
