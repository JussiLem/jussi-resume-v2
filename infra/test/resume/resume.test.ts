import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Resume } from '../../src/resume';

test('Snapshot', () => {
  const app = new App();
  const stack = new Resume(app, 'test', {
    certificateArn: 'arn:aws:acm:us-east-1:123456789012:certificate/feopoefwpfeopwfeopwfepoojggr-giogrge',
    domainName: 'DomainName',
    env: {
      account: '123456789012',
      region: 'eu-west-1',
    },
  });

  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});
