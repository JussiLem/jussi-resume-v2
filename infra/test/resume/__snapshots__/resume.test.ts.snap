// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Metadata": {
    "cdk_nag": {
      "rules_to_suppress": [
        {
          "id": "AwsSolutions-IAM4",
          "reason": "Change later to own policy",
        },
        {
          "id": "AwsSolutions-CFR3",
          "reason": "Save money and disable access logging",
        },
        {
          "id": "AwsSolutions-IAM5",
          "reason": "Change later to stop using wildcard",
        },
      ],
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": {
      "DependsOn": [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "de91205450df70f943fbde8ad6f4eec685d4c82c11ddec668199a8d2ec4d7065.zip",
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Ref": "NextJsStaticFilesDeploymentAwsCliLayer41531DE7",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "NextJsPublicAssetsS3Bucket06B7483E",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "NextJsPublicAssetsS3Bucket06B7483E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": [
          {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "40aa87cdf43c4095cec18bc443965f22ab2f8c1ace47e482a0ba4e35d83b0cc9.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "NextJsPublicAssetsS3Bucket06B7483E",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "NextJsDefaultCachePolicyB648229B": {
      "Properties": {
        "CachePolicyConfig": {
          "DefaultTTL": 2,
          "MaxTTL": 600,
          "MinTTL": 2,
          "Name": {
            "Fn::Join": [
              "",
              [
                "testNextJsDefaultCachePolicy8F4A0CF5-",
                {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "ParametersInCacheKeyAndForwardedToOrigin": {
            "CookiesConfig": {
              "CookieBehavior": "none",
            },
            "EnableAcceptEncodingBrotli": true,
            "EnableAcceptEncodingGzip": true,
            "HeadersConfig": {
              "HeaderBehavior": "none",
            },
            "QueryStringsConfig": {
              "QueryStringBehavior": "none",
            },
          },
        },
      },
      "Type": "AWS::CloudFront::CachePolicy",
    },
    "NextJsDistributionDB2DBAE8": {
      "Properties": {
        "DistributionConfig": {
          "Aliases": [
            "jussilemmetyinen.fi",
          ],
          "CacheBehaviors": [
            {
              "AllowedMethods": [
                "GET",
                "HEAD",
                "OPTIONS",
              ],
              "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
              "Compress": true,
              "OriginRequestPolicyId": "88a5eaf4-2fd4-4709-b370-b4c650ea3fcf",
              "PathPattern": "_next/*",
              "ResponseHeadersPolicyId": "67f7725c-6f97-4210-82d7-5512b31e9d03",
              "TargetOriginId": "testNextJsDistributionOrigin2325002D0",
              "ViewerProtocolPolicy": "redirect-to-https",
            },
          ],
          "DefaultCacheBehavior": {
            "AllowedMethods": [
              "GET",
              "HEAD",
              "OPTIONS",
            ],
            "CachePolicyId": {
              "Ref": "NextJsDefaultCachePolicyB648229B",
            },
            "Compress": true,
            "FunctionAssociations": [
              {
                "EventType": "viewer-request",
                "FunctionARN": {
                  "Fn::GetAtt": [
                    "NextJsFormatRequestLambda9F6134BC",
                    "FunctionARN",
                  ],
                },
              },
            ],
            "OriginRequestPolicyId": "88a5eaf4-2fd4-4709-b370-b4c650ea3fcf",
            "ResponseHeadersPolicyId": "67f7725c-6f97-4210-82d7-5512b31e9d03",
            "TargetOriginId": "testNextJsDistributionOrigin1A5CFF173",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "NextJsPublicAssetsS3Bucket06B7483E",
                  "RegionalDomainName",
                ],
              },
              "Id": "testNextJsDistributionOrigin1A5CFF173",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "NextJsDistributionOrigin1S3Origin02D90432",
                      },
                    ],
                  ],
                },
              },
            },
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "NextJsPublicAssetsS3Bucket06B7483E",
                  "RegionalDomainName",
                ],
              },
              "Id": "testNextJsDistributionOrigin2325002D0",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "NextJsDistributionOrigin2S3OriginA6D2CF24",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": {
            "AcmCertificateArn": "arn:aws:acm:us-east-1:123456789012:certificate/feopoefwpfeopwfeopwfepoojggr-giogrge",
            "MinimumProtocolVersion": "TLSv1.2_2021",
            "SslSupportMethod": "sni-only",
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "NextJsDistributionOrigin1S3Origin02D90432": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Identity for testNextJsDistributionOrigin1A5CFF173",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "NextJsDistributionOrigin2S3OriginA6D2CF24": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Identity for testNextJsDistributionOrigin2325002D0",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "NextJsFormatRequestLambda9F6134BC": {
      "Properties": {
        "AutoPublish": true,
        "FunctionCode": "/***
 * Append index.html or .html to the request path so that it matches the object inside S3
 * @param event
 */
function handler(event) {
    const {request} = event;
    const {uri} = request;
    if (uri === '/') {
        // turns "/" to "/index.html"
        request.uri += 'index.html';
    } else if (uri.endsWith('/')) {
        // turns "/foo/" to "/foo.html"
        request.uri = uri.slice(0, -1) + '.html';
    } else if (!uri.includes('.')) {
        // turns "/foo" to "/foo.html"
        request.uri += '.html';
    }
    return request;
}
",
        "FunctionConfig": {
          "Comment": "Formats path for S3",
          "Runtime": "cloudfront-js-1.0",
        },
        "Name": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::Region",
              },
              "testNextJsFormatRequestLambda2E245C40",
            ],
          ],
        },
      },
      "Type": "AWS::CloudFront::Function",
    },
    "NextJsPublicAssetsS3Bucket06B7483E": {
      "DeletionPolicy": "Delete",
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-S1",
              "reason": "For money saving",
            },
          ],
        },
      },
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
          {
            "Key": "aws-cdk:cr-owned:8f4c4aaf",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "NextJsPublicAssetsS3BucketAutoDeleteObjectsCustomResourceE9FF351F": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "NextJsPublicAssetsS3BucketPolicy47582B83",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "NextJsPublicAssetsS3Bucket06B7483E",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "NextJsPublicAssetsS3BucketPolicy47582B83": {
      "Properties": {
        "Bucket": {
          "Ref": "NextJsPublicAssetsS3Bucket06B7483E",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "NextJsPublicAssetsS3Bucket06B7483E",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "NextJsPublicAssetsS3Bucket06B7483E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "NextJsPublicAssetsS3Bucket06B7483E",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "NextJsPublicAssetsS3Bucket06B7483E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "NextJsDistributionOrigin1S3Origin02D90432",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "NextJsPublicAssetsS3Bucket06B7483E",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "NextJsDistributionOrigin2S3OriginA6D2CF24",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "NextJsPublicAssetsS3Bucket06B7483E",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "NextJsStaticFilesDeploymentAwsCliLayer41531DE7": {
      "Properties": {
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "159ab89ebd881b09102e294af7d28d2d93b12ed951fab089ece691faef1ffd44.zip",
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "NextJsStaticFilesDeploymentCustomResource7E320362": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "DestinationBucketName": {
          "Ref": "NextJsPublicAssetsS3Bucket06B7483E",
        },
        "Prune": false,
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": [
          {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
        ],
        "SourceObjectKeys": [
          "72aa5f3a737aec66e5ff385ecd3d381a65154419d842537917a6669c0ceed3f8.zip",
        ],
        "SystemMetadata": {
          "cache-control": "public, max-age=31536000, immutable",
        },
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
